### PR TITLE
feat: commit facts

### PR DESCRIPTION
**Pull request by anatolykislov**
Add new facts calculation:
- Average number of lines per commit (used for fun fact)
- Number of commits in repo (used for calculation of prev fun fact between all user's repos)
- Average length of line (used for fun fact)
- Number of locs for through all commits  (used for calculation of prev fun fact between all user's repos)
- Number of commits per number of lines (used for histogram in did you know section)

### CODE REVIEW

File: src/main/kotlin/app/hashers/FactHasher.kt [hunk 1]
```diff
@@ -105,4 +145,46 @@ class FactHasher(private val serverRepo: Repo = Repo(),
             Logger.info { "Sent ${facts.size} facts to server" }
         }
     }
+
+    private fun calcIncAvg(prev: Double, element: Double, count: Long):
```
Comment (asurkov) [inline src/main/kotlin/app/hashers/FactHasher.kt]: please give it a comment to make clear what the arguments are about

File: src/main/kotlin/app/hashers/FactHasher.kt [hunk 2]
```diff
@@ -105,4 +145,46 @@ class FactHasher(private val serverRepo: Repo = Repo(),
...
+        Double {
```
Comment (asurkov) [inline src/main/kotlin/app/hashers/FactHasher.kt]: btw, I find the style when arguments are on new lines more readable than putting a return type on a new line

File: src/main/kotlin/app/FactCodes.kt [hunk 3]
```diff
@@ -1,10 +1,20 @@
...
 package app
 
 object FactCodes {
     val COMMITS_DAY_WEEK = 1
     val COMMITS_DAY_TIME = 2
+    val COMMITS_LINE_NUM_AVG = 8
+    val COMMITS_NUM = 9
+    // Used for number of commits per number of lines in a commit histogram.
+    // Key should be number of lines. Value number of commits.
+    val COMMITS_NUM_PER_LINE_NUM = 12
```
Comment (asurkov) [inline src/main/kotlin/app/FactCodes.kt]: I find the name a bit confusing. Perhaps replacing 'per' to 'to' would make nicer? For example, COMMIT_COUNT_TO_LINE_COUNT. Also COMMITS -> COMMIT.

File: src/main/kotlin/app/hashers/FactHasher.kt [hunk 4]
```diff
@@ -19,56 +19,38 @@ import java.time.ZoneOffset
...
         for (author in emails) {
             fsDayWeek.put(author, Array(7) { 0 })
             fsDayTime.put(author, Array(24) { 0 })
             fsRepoDateStart.put(author, -1)
             fsRepoDateEnd.put(author, -1)
+            fsCommitLineNumAvg.put(author, 0.0)
+            fsCommitNum.put(author, 0)
+            fsLineLenAvg.put(author, 0.0)
+            fsLineNum.put(author, 0)
+            fsLinesPerCommits.put(author, Array(rehashes.size) {0})
```
[the remark this answers is not in the document]
Comment (anatolykislov) [reply src/main/kotlin/app/hashers/FactHasher.kt]: Good idea, I've add todo to future implementation.

File: src/main/kotlin/app/hashers/FactHasher.kt [hunk 5]
```diff
@@ -105,4 +141,55 @@ class FactHasher(private val serverRepo: Repo = Repo(),
             Logger.info { "Sent ${facts.size} facts to server" }
         }
     }
+
+    /**
+     * Used for incremental calculation of average of sequence.
```
Comment (asurkov) [inline src/main/kotlin/app/hashers/FactHasher.kt]: maybe: Computes the average of a numerical sequence

File: src/main/kotlin/app/FactCodes.kt [hunk 6]
```diff
@@ -1,10 +1,19 @@
...
+
 package app
 
 object FactCodes {
     val COMMITS_DAY_WEEK = 1
     val COMMITS_DAY_TIME = 2
+    val COMMITS_LINE_NUM_AVG = 8
+    val COMMITS_NUM = 9
+    // A map of line numbers to commits number. Used in a commit histogram.
+    val COMMITS_NUM_TO_LINE_NUM = 12
```
Comment (asurkov) [inline src/main/kotlin/app/FactCodes.kt]: i'm sort of concerned that commits_num has the plural form, but line_num has the single form.

File: src/main/kotlin/app/FactCodes.kt [hunk 7]
```diff
@@ -1,10 +1,19 @@
...
     val LINE_LONGEVITY = 3
     val LINE_LONGEVITY_REPO = 4
+    val LINE_LEN_AVG = 10
+    val LINE_NUM = 11
```
Comment (asurkov) [inline src/main/kotlin/app/FactCodes.kt]: it'd be really nice to document all of them.

Comment (anatolykislov) [reply src/main/kotlin/app/FactCodes.kt]: Agree

File: src/main/kotlin/app/FactCodes.kt (hunk 6 above)
Comment (anatolykislov) [reply src/main/kotlin/app/FactCodes.kt]: Fix

**Pull request merged**